File: CodeForces/Past Contest/Jiangsu Collegiate Programming Contest 2023/J.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class J{
   
   public static void main(String[] args)throws IOException{
      BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
      PrintWriter out = new PrintWriter(System.out);
      
      int t = Integer.parseInt(f.readLine());
      
      for(int q = 1; q <= t; q++){

         int n = Integer.parseInt(f.readLine());
      
         String[] array = new String[n];
         for(int k = 0; k < n; k++){
            array[k] = f.readLine();
         }
         
         int max = 0;
         for(int x1 = 0; x1 < n; x1++){
            for(int x2 = x1+1; x2 < n; x2++){
               int n1 = array[x1].length();
               int n2 = array[x2].length();
               
               int[][] dp = new int[n1+1][n2+1];
               for(int k = 0; k < n1; k++){
                  for(int j = 0; j < n2; j++){
                     if(array[x1].charAt(k) == array[x2].charAt(j)){
                        dp[k+1][j+1] = Math.max(dp[k+1][j+1],dp[k][j]+1);
                        max = Math.max(max,dp[k+1][j+1]);
                     }
                  }
               }
            }
         }
         
         out.println(max);

      }
      
      
      
      
      out.close();
   }
   
      
}

File: Kattis/Practice/Southeast USA Regional 2017-2018/J.kt
import java.io.*
import java.util.*
import kotlin.math.*

fun main(){
	val f = BufferedReader(InputStreamReader(System.`in`))

	val (n,m) = f.readLine().split(" ").map{it.toInt()}

	val gold = IntArray(n+1){0}
	val dec = IntArray(n+1){0}

	for(k in 1..n){
		val (g,d) = f.readLine().split(" ").map{it.toInt()}
		gold[k] = g
		dec[k] = d
	}

	val adj = Array(n+1){mutableListOf<Edge>()}

	for(k in 0 until m){
		val (a,b,t) = f.readLine().split(" ").map{it.toInt()}
		adj[a].add(Edge(b,t))
		adj[b].add(Edge(a,t))
	}

	val dp = Array(n+1){IntArray(1010){0}}

	dp[1][1] = gold[1]

	var answer = 0
	for(k in 1..1005){
		for(j in 1..n){
			if(dp[j][k] == 0) continue
			answer = max(dp[j][k],answer)
			for(e in adj[j]){
				if(e.t + k >= 1005) continue
				dp[e.to][e.t+k] = max(dp[e.to][e.t+k],dp[j][k] + max(0,gold[e.to]-(e.t+k-1)*dec[e.to]))
			}
		}
	}



	println(answer)

}

data class Edge(val to: Int, val t: Int)
